### PR TITLE
 Implemented additionally 2 behavioral patterns iterator and mediator

### PR DESCRIPTION
**Pull request by vantihovich**

### CODE REVIEW

File: patterns/behavioral/iterator/main.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		users: u.users,
+	}
+}
+
+func (u *userIterator) hasNext() bool {
+	return u.index < len(u.users)
+}
+
+func (u *userIterator) getNext() *user {
+	if u.hasNext() {
```
Comment (apaliavy) [inline patterns/behavioral/iterator/main.go:43]: happy-path your code: 
```
if !u.hasNext() {
    return nil 
}

... the rest of logic here
```
In that case it's easier to understand what's going on in conditional branches

File: patterns/behavioral/iterator/main.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		users: []*user{user1, user2, user3},
+	}
+
+	iterator := userCollection.createIterator()
+
+	for iterator.hasNext() {
+		user := iterator.getNext()
+		fmt.Printf("User is %+v\n", user)
+	}
+
```
Comment (apaliavy) [inline patterns/behavioral/iterator/main.go:78]: Blank line is not needed

Comment (vantihovich) [reply patterns/behavioral/iterator/main.go]: fixed

File: patterns/structural/facade/main.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+func (n *notification) sendWalletCreditNotification() {
+	fmt.Println("Sending notification on crediting wallet")
+}
+
+func (n *notification) sendWalletDebitNotification() {
+	fmt.Println("Sending notification on debiting wallet")
+}
+
+func main() {
+	fmt.Println()
```
[the remark this answers is not in the document]
Comment (vantihovich) [reply patterns/structural/facade/main.go:145]: just to make the output clear and readable

File: patterns/structural/facade/main.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	walletFacade := newWalletFacade("Ann", 1234)
+	fmt.Println()
```
[the remark this answers is not in the document]
Comment (vantihovich) [reply patterns/structural/facade/main.go:147]: just to make the output clear and readable

File: patterns/creational/prototype/main.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		copy := i.clone()
+		tempChildren = append(tempChildren, copy)
+	}
+
+	cloneFolder.children = tempChildren
+	return cloneFolder
+}
+
+func main() {
+
```
Comment (iamnotjustice) [inline patterns/creational/prototype/main.go:50]: Empty line is not needed here

File: patterns/behavioral/iterator/main.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		return nil
+	}
+
+	user := u.users[u.index]
+	u.index++
+	return user
+}
+
+func main() {
+
```
Comment (iamnotjustice) [inline patterns/behavioral/iterator/main.go:53]: Empty line again

File: patterns/structural/adapter/main.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	windowsMachine *windows
+}
+
+func (w *windowsAdapter) insertIntoLightningPort() {
+	fmt.Println("Adapter converts lightning signal to USB.")
+	w.windowsMachine.insertIntoUSBPort()
+}
+
+func main() {
+
```
Comment (iamnotjustice) [inline patterns/structural/adapter/main.go:38]: And again

**Review by iamnotjustice — APPROVED** [review]
Approved with some minor nitpicks ;)

File: patterns/structural/adapter/main.go (hunk 7 above)
Comment (vantihovich) [reply patterns/structural/adapter/main.go]: fixed

File: patterns/creational/prototype/main.go (hunk 5 above)
Comment (vantihovich) [reply patterns/creational/prototype/main.go]: fixed

**Pull request merged**